File: DesignPatterns/ProObjCDesignPatterns/Facade/Facade/ViewController.m
//
//  ViewController.m
//  Facade
//
//  Created by ShannonChen on 2017/11/11.
//  Copyright © 2017年 ShannonChen. All rights reserved.
//

#import "ViewController.h"
#import "CabDriver.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    /**
     外观模式：为系统中的一组接口提供一个统一的接口。外观定义一个高层接口，让子系统更易于使用。外观模式本质上就是抽象接口封装，隐藏复杂的细节。比如，对于开车的司机来说，他只需要提供目的地，并且再到达之后知道该花多少钱就行了，具体汽车怎么开过去的，不需要知道。
     
     */
    CabDriver *driver = [[CabDriver alloc] init];
    [driver driveToLocation:CGPointMake(100, 100) withCompletion:^(float price) {
        NSLog(@"共计%.2f元", price);
    }];
    
}





@end

File: ProtocolProgramming/Protocol/ImageEnity/ViewController.swift
//
//  ViewController.swift
//  ImageEnity
//
//  Created by ShannonChen on 2019/9/1.
//  Copyright © 2019 xianglongchen. All rights reserved.
//

/**
 Protocol Oriented Programming in Swift. by Károly Nyisztor
 https://www.pluralsight.com/guides/protocol-oriented-programming-in-swift
 
 类的方式 VS. POP 的方式
 
 */

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Test
        createImage()
        createMyImage()
        
    }

    func createImage() {
        
        let image = Image(name: "Pic", data: Data(repeating: 0, count: 100))
        print(image.base64Encoded)
        
        do {
            // persist image
            let documentDirectory = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor:nil, create:false)
            let imageURL = documentDirectory.appendingPathComponent("MyImage")
            try image.save(to: imageURL)
            print("Image saved successfully to path \(imageURL)")
            
            // load image from persistence
            let storedImage = try Image.init(name: "MyRestoredImage", contentsOf: imageURL)
            print("Image loaded successfully from path \(imageURL)")
        } catch {
            print(error)
        }
    }
    
    func createMyImage() {
        let image = MyImage(name: "Pic", data: Data(repeating: 0, count: 100))
        print(image.base64Encoded)
        
        do {
            // persist image
            let documentDirectory = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor:nil, create:false)
            let imageURL = documentDirectory.appendingPathComponent("MyImage")
            try image.save(to: imageURL)
            print("Image saved successfully to path \(imageURL)")
            
            // load image from persistence
            let storedImage = try MyImage.init(name: "MyRestoredImage", contentsOf: imageURL)
            print("Image loaded successfully from path \(imageURL)")
        } catch {
            print(error)
        }
    }

}

